011-kafka-canary.md

```diff
@@ -0,0 +1,40 @@
+# Strmixi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and continously producing and consuming events on the topics. 
+
+## Current situation
+
+Currently  Strimzi does not have any canary style feature. It does have readiness checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following:
+  
+  * Improved readiness check for Kafka clusters
+  * SLI(s) for service admins
+  * Metrics:
+    * End-to-end latency
+    * Succesfull/failed events
+    * Disk usage
+    * Disk errors
```

Comment:
How would you expect this info to be gathered?

Reply:
I think the point is that getting a successful acknowledgement to a produce request when acks=1 allows us to infer that the broker is able to write to the disk.

Reply:
@tombentley doesn't it just mean that the message was written in the page-cache so it's still in memory and not on the disk? The disk I/O error could come later when the operating system will do a flush, or?

Reply:
Well, for example with JBOD it would test only one disk out of N. So I think it can discover some disk errors as a side-effect. But it is not really a feature which should be advertised since it does not really check all disks and it really finds only disk errors which affects the particular consumer / producer. It doesn't check any system metrics or anything to find out about disk errors.

Reply:
Strictly speaking when changes actually make it to disk depends on a bunch of things including (but probably not limited to):

* The flush settings for the topic (in particular the max number of messages between flushes)
* The configuration of the OS
* And lower level things. (I believe it's not unusual/common for disks to have a battery backed memory cache which allows it to pretend that it wrote to disk while not actually, so even the kernel is being lied to). 

But it's a good point because the better we're tune the topic for disk error inference the less useful it is for metrics from which we can infer anything useful about performance. Kafka recommendation is to use replication rather than OS flush, of course.

Reply:
> Kafka recommendation is to use replication rather than OS flush, of course.

exactly! because data goes to the page cache and you really don't know when it will be written to disk. So in reality I don't think we can get a good info about disk error from producer.

Reply:
You'd just set `flush.messages = 1`

Reply:
I guess it's also worth discussing the merits of just checking `Admin.describeLogDirs()` on every reconciliation (maybe in KafkaRoller). That should tell you if a broker had observed an `IOException` while reading or writing from any segment in a given log directory. It would be less timely (the log directory could have been offline for as much as a reconciliation interval), and it wouldn't validate any of the rest of the broker (e.g. network stack etc). But it would be relatively simple.

Reply:
That says just to flush every message not ... when to flush. Of course, you know internal Kafka broker behavior better than me but how it reflects on the OS flush, just really calling "fsync" before the ack is sent? or sent the ack and then calling "fsync" ... so you can still lose the message and have I/O disk errors?

Reply:
Be careful with `flush.messages=1` ... while it is topic level config, I wouldn't be surprised of it to have potential to impact the overall performance and its predictability as well.